src/daemon/default_vm_image_vault.cpp

```diff
@@ -344,50 +344,44 @@ mp::VMImage mp::DefaultVMImageVault::fetch_image(const FetchType& fetch_type, co
-                                           kernel_info.initrd_location,
-                                           QString::fromStdString(id),
-                                           {},
-                                           0,
-                                           false};
-                    const auto image_filename = filename_for(image_url.path());
-                    // Attempt to make a sane directory name based on the filename of the image
-
-                    const auto image_dir_name =
-                        QString("%1-%2")
-                            .arg(image_filename.section(".", 0, image_filename.endsWith(".xz") ? -3 : -2))
-                            .arg(last_modified.toString("yyyyMMdd"));
-                    const auto image_dir = mp::utils::make_dir(images_dir, image_dir_name);
-
-                    // Had to use std::bind here to workaround the 5 allowable function arguments constraint of
-                    // QtConcurrent::run()
-                    future = QtConcurrent::run(std::bind(&DefaultVMImageVault::download_and_prepare_source_image, this,
-                                                         info, source_image, image_dir, fetch_type, prepare, monitor));
-
-                    in_progress_image_fetches[id] = future;
-                }
+                auto kernel_info = get_kernel_query_info(query.name);
+                const VMImageInfo info{{},
+                                       {},
+                                       {},
+                                       {},
+                                       true,
+                                       image_url.url(),
+                                       kernel_info.kernel_location,
+                                       kernel_info.initrd_location,
+                                       QString::fromStdString(id),
+                                       last_modified.toString(),
+                                       0,
+                                       false};
+                const auto image_filename = filename_for(image_url.path());
+                // Attempt to make a sane directory name based on the filename of the image
+
+                const auto image_dir_name =
+                    QString("%1-%2")
+                        .arg(image_filename.section(".", 0, image_filename.endsWith(".xz") ? -3 : -2))
+                        .arg(last_modified.toString("yyyyMMdd"));
+                const auto image_dir = mp::utils::make_dir(images_dir, image_dir_name);
+
+                // Had to use std::bind here to workaround the 5 allowable function arguments constraint of
+                // QtConcurrent::run()
+                future = QtConcurrent::run(std::bind(&DefaultVMImageVault::download_and_prepare_source_image, this,
+                                                     info, source_image, image_dir, fetch_type, prepare, monitor));
+
+                in_progress_image_fetches[id] = future;
```

Comment:
I'm not sure I follow what you mean, but again, no new code here.

Reply:
Here I was trying to make sure that a user moving multipass versions, from before to after this fix, would get the intended behavior, with no leftover state on disk or other unexpected effects.

Reply:
I don't see how this would happen.  If a user already has a cached http download type prepared image prior to this fix and then tries to launch an the same url and the lastUpdated http header has not changed, then line 345 should catch that and we won't get this point.  If the lastUpdated http header has changed, then it will download and prepare the updated image into the same directory and that will be used.  I hope this answers your question.

Reply:
OK, thanks. IIUC we would get into the second scenario here (dates not matching), because the first image would have been obtained before line 552 was added. But re-downloading/preparing would be the right thing to do.